Feature/vs 15 implement druid ip address filter extension

vivek807 opened the pull request.

vivek807 committed:
Initial commit for Druid IPRange Filter extension.

vivek807 committed:
Minor refactoring.

nozjkoitop commented on druid-filtering-ext/pom.xml:
```diff
@@ -0,0 +1,104 @@
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
```
Nit: This better to be consistent, we have junit4 in our parent pom already, i suggest moving everything to JUnit5

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/README.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+#### Parameter Descriptions
````
Parameter names are a bit confusing i suggest smth like
lower -> ipRangeLowerBound
upper -> ipRangeUpperBound
lowerOpen -> isLowerBoundOpen
upperOpen -> isUpperBoundOpen
ignoreVersionMismatch -> ignoreIPVersionMismatch

vivek807 replied:
It is something similar to `range` filter. I think we added it as part of design. Let me know we still need to change it.

nozjkoitop replied:
I would make field names more informative @cmick WDYT?

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPRange.java:
```diff
@@ -0,0 +1,104 @@
+        this.ipVersion = lower.getIPVersion();
+    }
+
+    public boolean contain(final IPAddress address) {
```
```suggestion
    public boolean contains(final IPAddress address) {
```

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPRange.java:
```diff
@@ -0,0 +1,104 @@
+            throw InvalidInput.exception("Invalid upper IP address");
+        }
+
+        if (lower.getIPVersion() != upper.getIPVersion()) {
```
```suggestion
       if (lower.getIPVersion().equals(upper.getIPVersion()))
```

vivek807 replied:
It is an enum, so equality will be same as equal

nozjkoitop replied:
Yeah, it'll just look a bit prettier this way, nothing else :)

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPRange.java:
```diff
@@ -0,0 +1,104 @@
+
+    @JsonCreator
+    public IPRange(String range) {
+        if (StringUtils.isEmpty(range)) {
```
To avoid white-spaces as well
```suggestion
        if (StringUtils.isBlank(range)) {
```

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPRange.java:
```diff
@@ -0,0 +1,104 @@
+            throw InvalidInput.exception("Range cannot be null or empty");
+        }
+
+        final String[] parts = range.split(SEPARATOR);
```
parts should be renamed to bounds IMO

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPBoundedRange.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    @VisibleForTesting
+    public boolean contain(IPAddress ipAddress, boolean ignoreVersionMismatch) {
```
```suggestion
    public boolean contains(IPAddress ipAddress, boolean ignoreVersionMismatch) {
```

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPBoundedRange.java:
```diff
@@ -0,0 +1,135 @@
+    @VisibleForTesting
+    public boolean contain(IPAddress ipAddress, boolean ignoreVersionMismatch) {
+        return ipAddress == null || ipAddress.getIPVersion() != this.ipVersion
+                ? ignoreVersionMismatch
```
Should it return true if versionMismatch is ignored?  Dont get the idea

vivek807 replied:
Yes, suppose if we wants to apply a range to IPv4 but also wants any IPv6 address.

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/common/PredicateFactory.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.ColumnValueSelector;
+
+public class PredicateFactory implements ValueMatcher {
```
Cant we use DruidObjectPredicate as in other filters? Seems to be a more native way

vivek807 replied:
I checked this one, but it seems like it is applicable to vectorized column, not sure if it is valid here

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/ip/range/FixedSetIPFilter.java:
```diff
@@ -0,0 +1,121 @@
+        final SortedSet<IPAddress> addresses = ranges.stream()
+                .map(this::map)
+                .filter(Objects::nonNull)
+                .sorted()
```
Why do we need to sort it?

vivek807 replied:
It is for performance, searching in sorted collection will be faster than non-sorted.

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/ip/range/impl/FixedSetIPFilterImpl.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @VisibleForTesting
+    public boolean contain(@NotNull final String addressStr) {
```
Please rename contain to contains everywhere

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/ip/range/impl/MultiRangeIPFilterImpl.java:
```diff
@@ -0,0 +1,117 @@
+    private final Set<IPRange> ipV6Ranges;
+    private final boolean ignoreVersionMismatch;
+
+    public MultiRangeIPFilterImpl(String column, Set<IPRange> ranges, @Nullable Boolean ignoreVersionMismatch) {
```
ignoreVersionMismatch is not nullable as primitive is sent there in MultiRangeIPFilter.toFilter, let's use primitive here as well

vivek807 replied:
Done

nozjkoitop commented on druid-filtering-ext/src/main/java/bi/deep/filtering/ip/range/impl/SingleTypeIPRangeFilterImpl.java:
```diff
@@ -0,0 +1,96 @@
+    private final IPBoundedRange boundedRange;
+    private final boolean ignoreVersionMismatch;
+
+    public SingleTypeIPRangeFilterImpl(String column, IPBoundedRange range, @Nullable Boolean ignoreVersionMismatch) {
```
Same for ignoreVersionMismatch

vivek807 replied:
Done

nozjkoitop reviewed (changes requested):
I would call it druid-ip-range-filter as druid-filtering-ext doesn't seem right
This extension also should be added to the [release workflow](https://github.com/deep-bi/deep-druid-extensions/blob/main/.github/workflows/release-flow.yml)

vivek807 commented:
> I would call it druid-ip-rane-filter as druid-filtering-ext doesn't seem right This extension also should be added to the [release workflow](https://github.com/deep-bi/deep-druid-extensions/blob/main/.github/workflows/release-flow.yml)

About name, I was thinking that we can add multiple filter here so named it like it. If it doesn't make much sense we can rename it

nozjkoitop commented:
> About name, I was thinking that we can add multiple filter here so named it like it. If it doesn't make much sense we can rename it

IDK whether we gonna add some new filters, but in this case we will be able to reorganize this. Cuz we can end up keeping it abstract for no reason.

BTW, should this extension be public?

vivek807 committed:
Addressed review comments.
Renamed `druid-filter-ext` to `druid-ip-range-filter`

vivek807 committed:
Addressed review comments.
Renamed `druid-filter-ext` to `druid-ip-range-filter`

vivek807 commented:
> > About name, I was thinking that we can add multiple filter here so named it like it. If it doesn't make much sense we can rename it
> 
> IDK whether we gonna add some new filters, but in this case we will be able to reorganize this. Cuz we can end up keeping it abstract for no reason.
> 
> BTW, should this extension be public?

No I dont think so it will be public.
Also I renamed it as suggested.

nozjkoitop commented:
> No I dont think so it will be public. Also I renamed it as suggested.

Current repo is a public one, if it should be private, than it should be moved to a separate repo

vivek807 commented:
> > No I dont think so it will be public. Also I renamed it as suggested.
> 
> Current repo is a public one, if it should be private, than it should be moved to a separate repo

Added new [PR](https://github.com/deep-bi/druid-ip-range-filter/pull/1) for new Private repo
